### ISSUE TITLE
`which` shows aliases as built-in commands

### ISSUE THREAD

**Issue by rgwood** [nushell/nushell#8577]
### Describe the bug

`which` incorrectly describes aliases as "Nushell built-in commands". I think it probably needs to be updated after @kubouch's recent alias changes.

### How to reproduce

```
> alias foo = print "foo!"
> foo
foo!
> which foo
╭───┬─────┬──────────────────────────┬──────────╮
│ # │ arg │           path           │ built-in │
├───┼─────┼──────────────────────────┼──────────┤
│ 0 │ foo │ Nushell built-in command │ true     │
╰───┴─────┴──────────────────────────┴──────────╯
```

### Expected behavior

I expect `which` to describe `foo` as an alias, and for the `built-in` column to be false.

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.77.2                                   |
| branch             | main                                     |
| commit_hash        | 66ad83c15ce0289a9901b5f618e9ea95ea763974 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| build_time         | 2023-03-22 18:59:19 -07:00               |
| build_rust_channel | debug                                    |
| features           | default, zip                             |
| installed_plugins  | inc                                      |

**Comment by kubouch** [comment]
That's because alias is a type of a command now. For someone attempting to look at it, the way to query aliases now is to query commands, then check `is_alias()` defined by the Command trait. A lot of the implementation details in the code referring to aliases, like `find_alias()`, refer to the old alias implementation and will be removed eventually.

**Comment by AyushSingh13** [comment]
I am happy to try working on this! 

Is @kubouch's change in question part of this PR: https://github.com/nushell/nushell/pull/8557

It would be my first contribution to nushell so I'm in the process of setting up my dev environment and can start exploring the code now! I've gone through the Review Process section of CONTRIBUTING.md but please let me know if I'm missing a step wrt assigning this to myself.

**Comment by AyushSingh13** [comment]
@kubouch - awesome, thank you! Will come back with loads of questions this weekend I'm sure :)

**Comment by AyushSingh13** [comment]
Not entirely related to this, so please let me know if it would make sense to open a new issue for this! 

# Overview
I'm running into an issue after installing `nushell` that's conflicting with my `tmux` setup. The kicker is that `brew uninstall nushell` doesn't fix the issue and my setup is still broken 😝.

# Details
Installing `nushell` via `brew install nushell` on an M1 Pro MacBook Pro, and reproduced on another M1 Max MacBook Pro, both running macOs Ventura 13.1 breaks `tmux` in the following ways.

After installing, `nu` works perfectly fine (i.e. I can access the REPL but `tmux new` and spawning new tmux windows using `prefix + c` breaks.

When starting a new tmux session using `tmux new`, the following occurs:
- `kitty` terminal
```bash
$ tmux new
[exited]
$ 1;4000;25c>|kitty(0.25.2)tmux>|kitty(0.25.2)>|kitty(0.25.2)>|kitty(0.25.2)>|kitty(0.25.2)
```
- `iterm2` terminal
```bash
$ tmux new
^[[>0;95;0c^[P>|iTerm2 3.4.19^[\[exited]
$ 0;95;0c>|iTerm2 3.4.19
```

When trying to open a new window in an existing tmux session using `prefix + c`, the new window opens and exits almost instantaneously.

All my existing tmux sessions and windows work perfectly fine - I'm just unable to spawn new ones!

# Investigation So Far
- checked the [Homebrew Formula](https://github.com/Homebrew/homebrew-core/blob/d2c5f6887fc14dd48506d08b52122ac08bb461e6/Formula/nushell.rb#L13)  for `nushell` to understand how this is installed/whether it edits any of my shell configs/other - the answer to this seems to be no as it's a `cargo install...`
- understanding what happens when opening a new tmux session/window. It looks like login scripts are executed but I can't find any clues in `/etc/profile` or `/etc/profile.d`

I feel like I might be missing something really obvious as I can't find anyone who has had similar issues so apologies in advance if this is the case! Please let me know if I should clarify any of the above/add any more details - thank you!

**Comment by rgwood** [comment]
^that doesn't seem related to this issue at all. You will likely have better luck filing a new issue or asking on Discord.

**Comment by AyushSingh13** [comment]
thank you - will create a new issue/ask on Discord! I have hidden the comment so it doesn't pollute this thread and will come back with specifics about this issue ASAP

**Comment by jooooscha** [comment]
Just to add my three cents: The `which` command in bash shows the definition of the alias. This is useful when you forget what exactly the alias does.

For example, for `alias hi=echo "hi"`, calling `which hi` would reveal the definition `echo "hi"`.

I think this is an important feature, and the nushell `which` command should show that, too.

**Comment by AyushSingh13** [comment]
@jooooscha - I think that's a great idea - I have started to explore the code so will report back on what changes are required for this :)

PS: for anyone that's interested, I fixed my `tmux` issue on both laptops by doing a `tmux kill-server`. I'm still not sure what the root cause was but the server seemed to have gone into some bad state and I haven't been able to reproduce it/link it explicitly back to `nushell`. 

Either way, will continue on with this issue finally! Apologies for the delay!

**Comment by AyushSingh13** [comment]
Update: I've spent some time exploring the code and I'm being a Rust ecosystem + language newbie at the same time so thank you for bearing with me.

I understand that `crates/nu-command/src/system/which_.rs` has a `which_single` function which eventually leads down to a `get_entry_in_commands` function which is responsible for returning `Nushell built-in command`/`Nushell custom command`. 

For some reason the call to check if something is a custom_command (`engine_state.get_decl(decl_id).is_custom_command()`) returns `false`. I also checked that `engine_state.aliases` is an empty vector even though I have defined `alias ll = ls -al` as my test alias.

So far I'm doing some newbie `println!` debugging but going to try and understand LLDB to see if I can do some more sophisticated debugging to understand the code path for adding an alias and then fetching an existing alias a bit better. Any advice on levelling up on the Rust side quicker is appreciated ofc :)

**Comment by kubouch** [comment]
`engine_state.aliases` is still the old alias implementation that will be deleted. Actually, most of the stuff referring to aliases in the engine is the old implementation and will be deleted. The current aliases are implemented as decls, but they should return `true` when you call `is_alias()` on them.

Inside `which_.rs` there is `get_entry_in_aliases()` which again refers to the old alias implementation and will be deleted. The correct way to query aliases would probably be to use the `get_entry_in_commands()` and then add another if-else branch there checking for `is_alias()`.

**Comment by AyushSingh13** [comment]
Thank you for your guidance and apologies for completely dropping the ball on this for so long! I have created this super simple draft PR here: https://github.com/nushell/nushell/pull/9580#pullrequestreview-1508714102

Just checking how the tests work and will hopefully be ready for review soon - thank you again for your patience!

Once this is done, I can an issue for implementing @jooooscha's idea of displaying alias definitions in `which` too.
